### ISSUE TITLE
app runs locally but not on azure

### ISSUE THREAD

**Issue by MarPeg53** [Azure-Samples/azure-search-openai-demo#951]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
> I made small adjustments in the backend and imported some modules. The app runs locally without problems but not on azure. I'm getting the error ModuleNotFoundError: No module named 'main'

### Any log messages given by the failure
>2023-11-13T15:00:51.048245136Z Documentation: http://aka.ms/webapp-linux
2023-11-13T15:00:51.048248537Z Python 3.11.4
2023-11-13T15:00:51.048252037Z Note: Any data outside '/home' is not persisted
2023-11-13T15:00:53.841410117Z Starting OpenBSD Secure Shell server: sshd.
2023-11-13T15:00:54.079495903Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-13T15:00:54.080804907Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-13T15:00:54.249114991Z Could not find build manifest file at '/home/site/wwwroot/oryx-manifest.toml'
2023-11-13T15:00:54.249849694Z Could not find operation ID in manifest. Generating an operation id...
2023-11-13T15:00:54.258543119Z Build Operation ID: 23ac646d-8475-407f-9678-e9a68c100d0c
2023-11-13T15:00:54.932481259Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-13T15:00:55.020314512Z Writing output script to '/opt/startup/startup.sh'
2023-11-13T15:00:55.144179167Z WARNING: Could not find virtual environment directory /home/site/wwwroot/antenv.
2023-11-13T15:00:55.151022287Z WARNING: Could not find package directory /home/site/wwwroot/__oryx_packages__.
2023-11-13T15:01:02.485772081Z [2023-11-13 15:01:02 +0000] [59] [INFO] Starting gunicorn 20.1.0
2023-11-13T15:01:02.507425354Z [2023-11-13 15:01:02 +0000] [59] [INFO] Listening at: http://0.0.0.0:8000 (59)
2023-11-13T15:01:02.507455954Z [2023-11-13 15:01:02 +0000] [59] [INFO] Using worker: sync
2023-11-13T15:01:02.720012673Z [2023-11-13 15:01:02 +0000] [65] [INFO] Booting worker with pid: 65
2023-11-13T15:01:02.807577969Z [2023-11-13 15:01:02 +0000] [65] [ERROR] Exception in worker process
2023-11-13T15:01:02.807624769Z Traceback (most recent call last):
2023-11-13T15:01:02.807630869Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
2023-11-13T15:01:02.807634869Z     worker.init_process()
2023-11-13T15:01:02.807638569Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
2023-11-13T15:01:02.807648869Z     self.load_wsgi()
2023-11-13T15:01:02.807653669Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
2023-11-13T15:01:02.807669369Z     self.wsgi = self.app.wsgi()
2023-11-13T15:01:02.807673869Z                 ^^^^^^^^^^^^^^^
2023-11-13T15:01:02.807677269Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
2023-11-13T15:01:02.807680969Z     self.callable = self.load()
2023-11-13T15:01:02.807684369Z                     ^^^^^^^^^^^
2023-11-13T15:01:02.807687769Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
2023-11-13T15:01:02.807691269Z     return self.load_wsgiapp()
2023-11-13T15:01:02.807694769Z            ^^^^^^^^^^^^^^^^^^^
2023-11-13T15:01:02.807698169Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
2023-11-13T15:01:02.807701769Z     return util.import_app(self.app_uri)
2023-11-13T15:01:02.807705169Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-13T15:01:02.807708669Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/util.py", line 359, in import_app
2023-11-13T15:01:02.807712269Z     mod = importlib.import_module(module)
2023-11-13T15:01:02.807715669Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-13T15:01:02.807719069Z   File "/opt/python/3/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-11-13T15:01:02.807777469Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-11-13T15:01:02.807786769Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-13T15:01:02.807790869Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-13T15:01:02.807794869Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-13T15:01:02.807798569Z   File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
2023-11-13T15:01:02.807802169Z ModuleNotFoundError: No module named 'main'
2023-11-13T15:01:02.818257005Z [2023-11-13 15:01:02 +0000] [65] [INFO] Worker exiting (pid: 65)
2023-11-13T15:01:04.082955771Z [2023-11-13 15:01:04 +0000] [59] [INFO] Shutting down: Master
2023-11-13T15:01:04.112193770Z [2023-11-13 15:01:04 +0000] [59] [INFO] Reason: Worker failed to boot.

### OS and Version?
Windows 11 Pro

### azd version?
azd version 1.4.3 (commit d165bd2de96dae75de57604c0d8a5553ae214618)

### logs deployment center
[logs deployment center.txt](https://github.com/Azure-Samples/azure-search-openai-demo/files/13336260/logs.deployment.center.txt)

**Comment by MarPeg53** [comment]
yes, I waited much longer than 10 minutes and it didn't help.

**Comment by MarPeg53** [comment]
yes, I waited much longer than 10 minutes and it didn't help.

**Comment by pamelafox** [comment]
Okay, well the error indicates that it can't find main.py, which means something went awry with the code upload, since app/backend/ should have a main.py. You could try re-running "azd deploy", and/or you could check the Deployment logs in the Azure Portal to see if it reported any issues with deployment. See first part of my blog post here:
https://blog.pamelafox.org/2023/06/tips-for-debugging-flask-deployments-to.html

**Comment by MarPeg53** [comment]
Thank you for you help Pamela. re-runnig with azd deploy didn't help and the logs also look okay from my point of view? Could it be that it has to do with the changes in the backend and the imported modules, even though it works locally? Any problems with Linux on azure and Windows locally?

**Comment by MarPeg53** [comment]
Thank you for your help Pamela. Re-running with azd deploy didn't help and the logs also look normal from my point of view? 
Could it be that it has to do with the changes in the backend and the imported modules, even though it works locally? Any problems with Linux on azure and Windows locally?

<img width="261" alt="image" src="https://github.com/Azure-Samples/azure-search-openai-demo/assets/100865916/c2006aed-4782-4782-9581-00ee67410803">


[Command oryx build.txt](https://github.com/Azure-Samples/azure-search-openai-demo/files/13348974/Command.oryx.build.txt)

**Comment by pamelafox** [comment]
Hm, the Oryx build logs look fine.

Your logs look like what I see when the app starts to restart the server before it has the code fully ready. It errors, but then once the code is ready, the next server restart works. I wonder if you can try Restart Server or running just azd deploy.

I'll share the logs from my most recent server start, it looks like yours at first and then fixes itself.

```
2023-11-14T22:46:50.590985786Z    _____                               
2023-11-14T22:46:50.591034989Z   /  _  \ __________ _________   ____  
2023-11-14T22:46:50.591046990Z  /  /_\  \\___   /  |  \_  __ \_/ __ \ 
2023-11-14T22:46:50.591051890Z /    |    \/    /|  |  /|  | \/\  ___/ 
2023-11-14T22:46:50.591056091Z \____|__  /_____ \____/ |__|    \___  >
2023-11-14T22:46:50.591061391Z         \/      \/                  \/ 
2023-11-14T22:46:50.591065491Z A P P   S E R V I C E   O N   L I N U X
2023-11-14T22:46:50.591069391Z 
2023-11-14T22:46:50.591073192Z Documentation: http://aka.ms/webapp-linux
2023-11-14T22:46:50.591077092Z Python 3.11.4
2023-11-14T22:46:50.591080992Z Note: Any data outside '/home' is not persisted
2023-11-14T22:47:00.677725263Z Starting OpenBSD Secure Shell server: sshd.
2023-11-14T22:47:01.212757873Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-14T22:47:02.321658713Z Starting periodic command scheduler: cron.
2023-11-14T22:47:02.321691315Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-14T22:47:02.812998063Z Could not find build manifest file at '/home/site/wwwroot/oryx-manifest.toml'
2023-11-14T22:47:02.814704385Z Could not find operation ID in manifest. Generating an operation id...
2023-11-14T22:47:02.821619280Z Build Operation ID: c5688804-09b6-47b3-ad6f-355838c172f3
2023-11-14T22:47:03.795591343Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-14T22:47:03.902461487Z Writing output script to '/opt/startup/startup.sh'
2023-11-14T22:47:04.066207297Z WARNING: Could not find virtual environment directory /home/site/wwwroot/antenv.
2023-11-14T22:47:04.066252400Z WARNING: Could not find package directory /home/site/wwwroot/__oryx_packages__.
2023-11-14T22:47:08.329477991Z [2023-11-14 22:47:08 +0000] [72] [INFO] Starting gunicorn 20.1.0
2023-11-14T22:47:08.338782656Z [2023-11-14 22:47:08 +0000] [72] [INFO] Listening at: http://0.0.0.0:8000 (72)
2023-11-14T22:47:08.338826059Z [2023-11-14 22:47:08 +0000] [72] [INFO] Using worker: sync
2023-11-14T22:47:08.416792928Z [2023-11-14 22:47:08 +0000] [73] [INFO] Booting worker with pid: 73
2023-11-14T22:47:08.682769626Z [2023-11-14 22:47:08 +0000] [73] [ERROR] Exception in worker process
2023-11-14T22:47:08.682839031Z Traceback (most recent call last):
2023-11-14T22:47:08.682846231Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
2023-11-14T22:47:08.682852531Z     worker.init_process()
2023-11-14T22:47:08.682857532Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
2023-11-14T22:47:08.682862732Z     self.load_wsgi()
2023-11-14T22:47:08.682885634Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
2023-11-14T22:47:08.682891134Z     self.wsgi = self.app.wsgi()
2023-11-14T22:47:08.682895735Z                 ^^^^^^^^^^^^^^^
2023-11-14T22:47:08.682900435Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
2023-11-14T22:47:08.682905235Z     self.callable = self.load()
2023-11-14T22:47:08.682909836Z                     ^^^^^^^^^^^
2023-11-14T22:47:08.682914536Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
2023-11-14T22:47:08.682919636Z     return self.load_wsgiapp()
2023-11-14T22:47:08.682924337Z            ^^^^^^^^^^^^^^^^^^^
2023-11-14T22:47:08.682928937Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
2023-11-14T22:47:08.682933637Z     return util.import_app(self.app_uri)
2023-11-14T22:47:08.682938238Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-14T22:47:08.682942838Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/util.py", line 359, in import_app
2023-11-14T22:47:08.682947738Z     mod = importlib.import_module(module)
2023-11-14T22:47:08.682952239Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-14T22:47:08.682956839Z   File "/opt/python/3/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-11-14T22:47:08.682961539Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-11-14T22:47:08.682966240Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-14T22:47:08.682971240Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-14T22:47:08.682976540Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-14T22:47:08.682981541Z   File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
2023-11-14T22:47:08.682986341Z ModuleNotFoundError: No module named 'main'
2023-11-14T22:47:08.691051117Z [2023-11-14 22:47:08 +0000] [73] [INFO] Worker exiting (pid: 73)
2023-11-14T22:47:09.443242935Z [2023-11-14 22:47:09 +0000] [72] [INFO] Shutting down: Master
2023-11-14T22:47:09.444523826Z [2023-11-14 22:47:09 +0000] [72] [INFO] Reason: Worker failed to boot.

2023-11-14T22:52:03.964739739Z    _____                               
2023-11-14T22:52:03.964788142Z   /  _  \ __________ _________   ____  
2023-11-14T22:52:03.964793742Z  /  /_\  \\___   /  |  \_  __ \_/ __ \ 
2023-11-14T22:52:03.964798242Z /    |    \/    /|  |  /|  | \/\  ___/ 
2023-11-14T22:52:03.964802542Z \____|__  /_____ \____/ |__|    \___  >
2023-11-14T22:52:03.964806943Z         \/      \/                  \/ 
2023-11-14T22:52:03.964811243Z A P P   S E R V I C E   O N   L I N U X
2023-11-14T22:52:03.964815343Z 
2023-11-14T22:52:03.964819243Z Documentation: http://aka.ms/webapp-linux
2023-11-14T22:52:03.964823343Z Python 3.11.4
2023-11-14T22:52:03.964827344Z Note: Any data outside '/home' is not persisted
2023-11-14T22:52:06.786887176Z Starting OpenBSD Secure Shell server: sshd.
2023-11-14T22:52:07.416817170Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-14T22:52:08.266259617Z Starting periodic command scheduler: cron.
2023-11-14T22:52:08.267829604Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-14T22:52:08.383778325Z Found build manifest file at '/home/site/wwwroot/oryx-manifest.toml'. Deserializing it...
2023-11-14T22:52:08.386990003Z Build Operation ID: 1cfa1d0420698611
2023-11-14T22:52:08.388924710Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-14T22:52:08.394796935Z Output is compressed. Extracting it...
2023-11-14T22:52:08.495341403Z Extracting '/home/site/wwwroot/output.tar.gz' to directory '/tmp/8dbe562e63641aa'...
2023-11-14T22:52:38.954832207Z App path is set to '/tmp/8dbe562e63641aa'
2023-11-14T22:52:39.082043924Z Writing output script to '/opt/startup/startup.sh'
2023-11-14T22:52:39.212233861Z Using packages from virtual environment antenv located at /tmp/8dbe562e63641aa/antenv.
2023-11-14T22:52:39.212274963Z Updated PYTHONPATH to '/opt/startup/app_logs:/tmp/8dbe562e63641aa/antenv/lib/python3.11/site-packages'
2023-11-14T22:52:41.003229677Z [2023-11-14 22:52:41 +0000] [74] [INFO] Starting gunicorn 20.1.0
2023-11-14T22:52:41.005456895Z [2023-11-14 22:52:41 +0000] [74] [INFO] Listening at: http://0.0.0.0:8000 (74)
2023-11-14T22:52:41.011535719Z [2023-11-14 22:52:41 +0000] [74] [INFO] Using worker: uvicorn.workers.UvicornWorker
2023-11-14T22:52:41.033938612Z [2023-11-14 22:52:41 +0000] [75] [INFO] Booting worker with pid: 75
2023-11-14T22:52:41.102637171Z [2023-11-14 22:52:41 +0000] [76] [INFO] Booting worker with pid: 76
2023-11-14T22:52:41.163480212Z [2023-11-14 22:52:41 +0000] [77] [INFO] Booting worker with pid: 77
2023-11-14T22:53:13.740452991Z [2023-11-14 22:53:13 +0000] [76] [INFO] Started server process [76]
2023-11-14T22:53:13.741783053Z [2023-11-14 22:53:13 +0000] [76] [INFO] Waiting for application startup.
2023-11-14T22:53:13.751407099Z [2023-11-14 22:53:13 +0000] [75] [INFO] Started server process [75]
2023-11-14T22:53:13.752934270Z [2023-11-14 22:53:13 +0000] [75] [INFO] Waiting for application startup.
2023-11-14T22:53:13.761449965Z WARNING:azure.identity.aio._credentials.environment:Incomplete environment configuration. These variables are set: AZURE_TENANT_ID
2023-11-14T22:53:13.770482684Z WARNING:azure.identity.aio._credentials.environment:Incomplete environment configuration. These variables are set: AZURE_TENANT_ID
2023-11-14T22:53:13.780911967Z [2023-11-14 22:53:13 +0000] [77] [INFO] Started server process [77]
2023-11-14T22:53:13.790122794Z [2023-11-14 22:53:13 +0000] [77] [INFO] Waiting for application startup.
2023-11-14T22:53:13.793155135Z WARNING:azure.identity.aio._credentials.environment:Incomplete environment configuration. These variables are set: AZURE_TENANT_ID
2023-11-14T22:53:17.705159396Z [2023-11-14 22:53:17 +0000] [76] [INFO] Application startup complete.
2023-11-14T22:53:17.714583598Z [2023-11-14 22:53:17 +0000] [75] [INFO] Application startup complete.
2023-11-14T22:53:17.736555867Z [2023-11-14 22:53:17 +0000] [77] [INFO] Application startup complete.
```

**Comment by MarPeg53** [comment]
Rerunning server or deploy again results always in the same error.  Do you have any other ideas about what could be the reason?

**Comment by pamelafox** [comment]
@MarPeg53 Are you able to SSH into the server from the Portal? If you can, then you can poke around and see what files are actually there. If you can't SSH due to the issue with starting the server, you could temporarily override the startup command to something that should always work (an empty string might work), and then SSH.

**Comment by Cornelius000** [comment]
I am currently having a similar problem but the logs say that the packages are not installed:
`2023-11-21T12:13:40.944223980Z    _____                               
2023-11-21T12:13:40.944257682Z   /  _  \ __________ _________   ____  
2023-11-21T12:13:40.944264083Z  /  /_\  \\___   /  |  \_  __ \_/ __ \ 
2023-11-21T12:13:40.944269383Z /    |    \/    /|  |  /|  | \/\  ___/ 
2023-11-21T12:13:40.944274383Z \____|__  /_____ \____/ |__|    \___  >
2023-11-21T12:13:40.944279684Z         \/      \/                  \/ 
2023-11-21T12:13:40.944284584Z A P P   S E R V I C E   O N   L I N U X
2023-11-21T12:13:40.944289784Z 
2023-11-21T12:13:40.944294284Z Documentation: http://aka.ms/webapp-linux
2023-11-21T12:13:40.944298985Z Python 3.11.4
2023-11-21T12:13:40.944303585Z Note: Any data outside '/home' is not persisted
2023-11-21T12:13:44.612375938Z Starting OpenBSD Secure Shell server: sshd.
2023-11-21T12:13:44.881892812Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-21T12:13:44.901114330Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-21T12:13:44.986868417Z Could not find build manifest file at '/home/site/wwwroot/oryx-manifest.toml'
2023-11-21T12:13:44.986905319Z Could not find operation ID in manifest. Generating an operation id...
2023-11-21T12:13:45.010308680Z Build Operation ID: 24bd0d0e-84a0-4b5b-8267-40f01cb73565
2023-11-21T12:13:46.225849324Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-21T12:13:46.275440404Z Writing output script to '/opt/startup/startup.sh'
2023-11-21T12:13:46.423344996Z WARNING: Could not find virtual environment directory /home/site/wwwroot/antenv.
2023-11-21T12:13:46.423373198Z WARNING: Could not find package directory /home/site/wwwroot/__oryx_packages__.
2023-11-21T12:13:49.039852182Z [2023-11-21 12:13:49 +0000] [64] [INFO] Starting gunicorn 20.1.0
2023-11-21T12:13:49.040867540Z [2023-11-21 12:13:49 +0000] [64] [INFO] Listening at: http://0.0.0.0:8000 (64)
2023-11-21T12:13:49.059901944Z [2023-11-21 12:13:49 +0000] [64] [INFO] Using worker: sync
2023-11-21T12:13:49.087688656Z [2023-11-21 12:13:49 +0000] [69] [INFO] Booting worker with pid: 69
2023-11-21T12:13:49.157315593Z [2023-11-21 12:13:49 +0000] [69] [ERROR] Exception in worker process
2023-11-21T12:13:49.157352495Z Traceback (most recent call last):
2023-11-21T12:13:49.157358796Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
2023-11-21T12:13:49.157363796Z     worker.init_process()
2023-11-21T12:13:49.157368196Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
2023-11-21T12:13:49.157374196Z     self.load_wsgi()
2023-11-21T12:13:49.157378697Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
2023-11-21T12:13:49.157397198Z     self.wsgi = self.app.wsgi()
2023-11-21T12:13:49.157401798Z                 ^^^^^^^^^^^^^^^
2023-11-21T12:13:49.157406298Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
2023-11-21T12:13:49.157410999Z     self.callable = self.load()
2023-11-21T12:13:49.157415299Z                     ^^^^^^^^^^^
2023-11-21T12:13:49.157419599Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
2023-11-21T12:13:49.157424299Z     return self.load_wsgiapp()
2023-11-21T12:13:49.157428700Z            ^^^^^^^^^^^^^^^^^^^
2023-11-21T12:13:49.157433100Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
2023-11-21T12:13:49.157437700Z     return util.import_app(self.app_uri)
2023-11-21T12:13:49.157442600Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:13:49.157447001Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/util.py", line 359, in import_app
2023-11-21T12:13:49.157451601Z     mod = importlib.import_module(module)
2023-11-21T12:13:49.157456001Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:13:49.157460501Z   File "/opt/python/3/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-11-21T12:13:49.157465202Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-11-21T12:13:49.157469602Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:13:49.157474002Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-21T12:13:49.157479003Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-21T12:13:49.157483603Z   File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
2023-11-21T12:13:49.157488303Z ModuleNotFoundError: No module named 'main'
2023-11-21T12:13:49.200106874Z [2023-11-21 12:13:49 +0000] [69] [INFO] Worker exiting (pid: 69)
2023-11-21T12:13:49.555368275Z [2023-11-21 12:13:49 +0000] [64] [INFO] Shutting down: Master
2023-11-21T12:13:49.555401777Z [2023-11-21 12:13:49 +0000] [64] [INFO] Reason: Worker failed to boot.

2023-11-21T12:14:34.911671056Z    _____                               
2023-11-21T12:14:34.911742759Z   /  _  \ __________ _________   ____  
2023-11-21T12:14:34.911756360Z  /  /_\  \\___   /  |  \_  __ \_/ __ \ 
2023-11-21T12:14:34.911767760Z /    |    \/    /|  |  /|  | \/\  ___/ 
2023-11-21T12:14:34.911778161Z \____|__  /_____ \____/ |__|    \___  >
2023-11-21T12:14:34.911788861Z         \/      \/                  \/ 
2023-11-21T12:14:34.911798962Z A P P   S E R V I C E   O N   L I N U X
2023-11-21T12:14:34.911809362Z 
2023-11-21T12:14:34.911818763Z Documentation: http://aka.ms/webapp-linux
2023-11-21T12:14:34.911828463Z Python 3.11.4
2023-11-21T12:14:34.911838164Z Note: Any data outside '/home' is not persisted
2023-11-21T12:14:43.562010131Z Starting OpenBSD Secure Shell server: sshd.
2023-11-21T12:14:44.077131166Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-21T12:14:44.093301364Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-21T12:14:44.971639967Z Could not find build manifest file at '/home/site/wwwroot/oryx-manifest.toml'
2023-11-21T12:14:44.971675969Z Could not find operation ID in manifest. Generating an operation id...
2023-11-21T12:14:45.017238300Z Build Operation ID: 9ce734cb-7615-46f0-b0cc-9baa6f790352
2023-11-21T12:14:47.302215042Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-21T12:14:47.375969934Z Writing output script to '/opt/startup/startup.sh'
2023-11-21T12:14:47.557622110Z WARNING: Could not find virtual environment directory /home/site/wwwroot/antenv.
2023-11-21T12:14:47.557658712Z WARNING: Could not find package directory /home/site/wwwroot/__oryx_packages__.
2023-11-21T12:14:53.181835041Z [2023-11-21 12:14:53 +0000] [65] [INFO] Starting gunicorn 20.1.0
2023-11-21T12:14:53.187295040Z [2023-11-21 12:14:53 +0000] [65] [INFO] Listening at: http://0.0.0.0:8000 (65)
2023-11-21T12:14:53.187771066Z [2023-11-21 12:14:53 +0000] [65] [INFO] Using worker: sync
2023-11-21T12:14:53.214569833Z [2023-11-21 12:14:53 +0000] [70] [INFO] Booting worker with pid: 70
2023-11-21T12:14:53.446951556Z [2023-11-21 12:14:53 +0000] [70] [ERROR] Exception in worker process
2023-11-21T12:14:53.446985358Z Traceback (most recent call last):
2023-11-21T12:14:53.446992058Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
2023-11-21T12:14:53.446997659Z     worker.init_process()
2023-11-21T12:14:53.447002759Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 134, in init_process
2023-11-21T12:14:53.447008059Z     self.load_wsgi()
2023-11-21T12:14:53.447019160Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
2023-11-21T12:14:53.447035761Z     self.wsgi = self.app.wsgi()
2023-11-21T12:14:53.447040761Z                 ^^^^^^^^^^^^^^^
2023-11-21T12:14:53.447045361Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/base.py", line 67, in wsgi
2023-11-21T12:14:53.447050162Z     self.callable = self.load()
2023-11-21T12:14:53.447058062Z                     ^^^^^^^^^^^
2023-11-21T12:14:53.447062762Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
2023-11-21T12:14:53.447067463Z     return self.load_wsgiapp()
2023-11-21T12:14:53.447152667Z            ^^^^^^^^^^^^^^^^^^^
2023-11-21T12:14:53.447160568Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
2023-11-21T12:14:53.447165668Z     return util.import_app(self.app_uri)
2023-11-21T12:14:53.447170368Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:14:53.447174868Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/util.py", line 359, in import_app
2023-11-21T12:14:53.447179769Z     mod = importlib.import_module(module)
2023-11-21T12:14:53.447184469Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:14:53.447189069Z   File "/opt/python/3/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-11-21T12:14:53.447193869Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-11-21T12:14:53.447198370Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:14:53.447202970Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-21T12:14:53.447208070Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-21T12:14:53.447212871Z   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
2023-11-21T12:14:53.447226471Z   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
2023-11-21T12:14:53.447231272Z   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
2023-11-21T12:14:53.447235872Z   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
2023-11-21T12:14:53.447240572Z   File "/home/site/wwwroot/main.py", line 1, in <module>
2023-11-21T12:14:53.447245472Z     from app import create_app
2023-11-21T12:14:53.447249873Z   File "/home/site/wwwroot/app.py", line 10, in <module>
2023-11-21T12:14:53.447254673Z     import aiohttp
2023-11-21T12:14:53.447260173Z ModuleNotFoundError: No module named 'aiohttp'
2023-11-21T12:14:53.458161170Z [2023-11-21 12:14:53 +0000] [70] [INFO] Worker exiting (pid: 70)
2023-11-21T12:14:54.116269499Z [2023-11-21 12:14:54 +0000] [65] [INFO] Shutting down: Master
2023-11-21T12:14:54.143887710Z [2023-11-21 12:14:54 +0000] [65] [INFO] Reason: Worker failed to boot.

2023-11-21T12:15:33.005825214Z    _____                               
2023-11-21T12:15:33.005865816Z   /  _  \ __________ _________   ____  
2023-11-21T12:15:33.005872917Z  /  /_\  \\___   /  |  \_  __ \_/ __ \ 
2023-11-21T12:15:33.005878717Z /    |    \/    /|  |  /|  | \/\  ___/ 
2023-11-21T12:15:33.005883917Z \____|__  /_____ \____/ |__|    \___  >
2023-11-21T12:15:33.005889618Z         \/      \/                  \/ 
2023-11-21T12:15:33.005894718Z A P P   S E R V I C E   O N   L I N U X
2023-11-21T12:15:33.005899618Z 
2023-11-21T12:15:33.005904219Z Documentation: http://aka.ms/webapp-linux
2023-11-21T12:15:33.005908919Z Python 3.11.4
2023-11-21T12:15:33.005913719Z Note: Any data outside '/home' is not persisted
2023-11-21T12:15:37.514308173Z Starting OpenBSD Secure Shell server: sshd.
2023-11-21T12:15:37.704593824Z Site's appCommandLine: python3 -m gunicorn main:app
2023-11-21T12:15:37.711393186Z Launching oryx with: create-script -appPath /home/site/wwwroot -output /opt/startup/startup.sh -virtualEnvName antenv -defaultApp /opt/defaultsite -userStartupCommand 'python3 -m gunicorn main:app'
2023-11-21T12:15:37.790486506Z Could not find build manifest file at '/home/site/wwwroot/oryx-manifest.toml'
2023-11-21T12:15:37.790519907Z Could not find operation ID in manifest. Generating an operation id...
2023-11-21T12:15:37.790526608Z Build Operation ID: 6ddbc81d-7087-49e5-ad30-9af4ef603f7f
2023-11-21T12:15:37.982886469Z Oryx Version: 0.2.20230707.1, Commit: 0bd28e69919b5e8beba451e8677e3345f0be8361, ReleaseTagName: 20230707.1
2023-11-21T12:15:38.011816212Z Writing output script to '/opt/startup/startup.sh'
2023-11-21T12:15:38.138021440Z WARNING: Could not find virtual environment directory /home/site/wwwroot/antenv.
2023-11-21T12:15:38.140298162Z WARNING: Could not find package directory /home/site/wwwroot/__oryx_packages__.
2023-11-21T12:15:40.910241867Z 
2023-11-21T12:15:40.910306071Z Error: class uri 'uvicorn.workers.UvicornWorker' invalid or not found: 
2023-11-21T12:15:40.910313371Z 
2023-11-21T12:15:40.910318071Z [Traceback (most recent call last):
2023-11-21T12:15:40.910322772Z   File "/opt/python/3/lib/python3.11/site-packages/gunicorn/util.py", line 99, in load_class
2023-11-21T12:15:40.910327372Z     mod = importlib.import_module('.'.join(components))
2023-11-21T12:15:40.910331872Z           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:15:40.910336272Z   File "/opt/python/3/lib/python3.11/importlib/__init__.py", line 126, in import_module
2023-11-21T12:15:40.910347573Z     return _bootstrap._gcd_import(name[level:], package, level)
2023-11-21T12:15:40.910351973Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-11-21T12:15:40.910356373Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-21T12:15:40.910361474Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-21T12:15:40.910375274Z   File "<frozen importlib._bootstrap>", line 1126, in _find_and_load_unlocked
2023-11-21T12:15:40.910379975Z   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
2023-11-21T12:15:40.910384575Z   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2023-11-21T12:15:40.910389075Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2023-11-21T12:15:40.910393575Z   File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
2023-11-21T12:15:40.910398076Z ModuleNotFoundError: No module named 'uvicorn'
2023-11-21T12:15:40.910402576Z ]
2023-11-21T12:15:40.964932080Z `

**Comment by pamelafox** [comment]
The error "No module named 'main'" generally means that the App Service does not have your application code (it cannot find main.py, which should be in the backend folder that gets uploaded). It's very common for that to be the case a few minutes after deploy while the server is still warming up. However, if you still see it, then I recommend checking the Deployment logs to make sure your last deployment was successful. You may need to do another deployment if the last one failed.

More details in my post at https://blog.pamelafox.org/2023/06/tips-for-debugging-flask-deployments-to.html

**Comment by chazarabriseis** [comment]
@pamelafox checking the deployment logs was a great hint! Thank you! I reversed the last changes and now it works again. Need to check what went wrong there...

**Comment by pamelafox** [comment]
Thanks for reporting this. Since there hasn’t been any activity since early 2024 and I don’t have enough to investigate further, I’m going to close this out.

If you’re still seeing “ModuleNotFoundError: No module named 'main'” after deploying, please open a new issue with deployment logs, the exact azd command you ran (from the repo root), and whether waiting a few minutes for warm-up or re-running azd down/up helps. Also check Deployment Center/SSH to confirm app/backend is present in /home/site/wwwroot. 

For a more reliable host, consider deploying to Azure Container Apps, which is now the default for reliability reasons.
